finality-aleph/src/sync/forest/mod.rs

```diff
@@ -101,26 +119,65 @@ impl<I: PeerId, J: Justification> VertexWithChildren<I, J> {
 // ever need worst case scenario.
 const MAX_DEPTH: u32 = 1800;
 
-pub struct Forest<I: PeerId, J: Justification> {
+pub struct Forest<B, I, J, CS>
+where
+    B: Block,
+    I: PeerId,
+    J: Justification<Header = B::Header>,
+    CS: ChainStatus<B, J>,
+{
     vertices: HashMap<BlockIdFor<J>, VertexWithChildren<I, J>>,
     highest_justified: BlockIdFor<J>,
     justified_blocks: HashMap<u32, BlockIdFor<J>>,
     root_id: BlockIdFor<J>,
     root_children: HashSet<BlockIdFor<J>>,
     compost_bin: HashSet<BlockIdFor<J>>,
+    _phantom: PhantomData<(B, CS)>,
 }
 
-impl<I: PeerId, J: Justification> Forest<I, J> {
-    pub fn new(highest_justified: BlockIdFor<J>) -> Self {
-        let root_id = highest_justified.clone();
-        Self {
+type Edge<J> = (BlockIdFor<J>, BlockIdFor<J>);
+
+impl<B, I, J, CS> Forest<B, I, J, CS>
+where
+    B: Block,
+    I: PeerId,
+    J: Justification<Header = B::Header>,
+    CS: ChainStatus<B, J>,
+{
+    pub fn new(chain_status: &CS) -> Result<Self, Error<B, J, CS>> {
+        let top_finalized = chain_status
+            .top_finalized()
+            .map_err(Error::ChainStatus)?
+            .header()
+            .id();
+        let mut forest = Self {
             vertices: HashMap::new(),
-            highest_justified,
+            highest_justified: top_finalized.clone(),
             justified_blocks: HashMap::new(),
-            root_id,
+            root_id: top_finalized.clone(),
             root_children: HashSet::new(),
             compost_bin: HashSet::new(),
+            _phantom: PhantomData,
+        };
+        let mut deque = VecDeque::from([top_finalized]);
```

Comment:
pls extract this logic to fn

Reply:
why? this is basically the only interesting thing happening inside `new`

Reply:
still in `new`, but looks much better now